### PR TITLE
BGA-13 Send token

### PR DESCRIPTION
**Pull request by gonzaalvarez70**
Initial implementation for sendMultisigToken transaction in account-lib for Celo.

### CODE REVIEW

File: src/coin/eth/index.ts [hunk 1]
```diff
@@ -2,6 +2,8 @@ export { KeyPair } from './keyPair';
 export { Transaction } from './transaction';
 export { TransactionBuilder } from './transactionBuilder';
 export { TransferBuilder } from './transferBuilder';
```
Comment (marktoda) [inline src/coin/eth/index.ts:4]: Do we still need to export `TransferBuilder` now that it is sort-of abstract?

File: src/coin/eth/transactionBuilder.ts [hunk 2]
```diff
@@ -128,6 +132,13 @@ export class TransactionBuilder extends BaseTransactionBuilder {
           this.owner(element);
         });
         break;
+      case TransactionType.Send:
+        if (transactionJson.to === undefined) {
+          throw new BuildTransactionError('Undefined recipient address');
+        }
+        this._contractAddress = transactionJson.to;
+        this._transfer = new TransferFundsBuilder(transactionJson.data);
+        break;
```
Comment (marktoda) [inline src/coin/eth/transactionBuilder.ts:141]: Should you add a case here for `SendToken` as well?

File: src/coin/eth/transferFundsBuilder.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
+import ethUtil from 'ethereumjs-util';
+import EthereumAbi from 'ethereumjs-abi';
+import BigNumber from 'bignumber.js';
+import { BuildTransactionError } from '../baseCoin/errors';
+import { sendMultiSigData } from './utils';
+import { sendMultisigMethodId } from './walletUtil';
+import { TransferBuilder } from './';
+
+/** ETH transfer builder */
+export class TransferFundsBuilder extends TransferBuilder {
```
Comment (marktoda) [inline src/coin/eth/transferFundsBuilder.ts:10]: Instead of having two entirely different builder classes for transferring native assets and tokens, what if we just kept the one `TransferBuilder` and added a `coin` field (and builder function) that:
- defaults to the native currency (cgld, rbtc, etc)
- if the user inputs anything other than the native currency, it checks it against @bitgo/statics coin list: https://github.com/BitGo/BitGoJS/tree/master/modules/statics
- using the @bitgo/statics coin details, it can determine if the given asset is a token -- if it's a token then we can make the necessary changes to the build flow: use the token `operation hash` and the token function call.

What do you think? I think that's a simpler user API and also probably simpler code. cc @CapnMigraine what are you guys doing for XTZ in this situation?

File: src/coin/eth/utils.ts [hunk 4]
```diff
@@ -98,6 +90,32 @@ export function sendMultiSigData(
...
+export function sendMultiSigTokenData(
+  to: string,
+  value: number,
+  tokenContractAddress: string,
+  expireTime: number,
+  sequenceId: number,
+  signature: string,
+): string {
+  const params = [to, value, tokenContractAddress, expireTime, sequenceId, toBuffer(signature)];
+  const types = ['address', 'uint', 'address', 'uint', 'uint', 'bytes'];
```
Comment (marktoda) [inline src/coin/eth/utils.ts:113]: can you save the types in `walletUtil` like we have done for the constructor args

File: src/coin/eth/index.ts (hunk 1 above)
Comment (gonzaalvarez70) [reply src/coin/eth/index.ts:4]: You are right we don't need it anymore.

File: src/coin/eth/transactionBuilder.ts (hunk 2 above)
Comment (gonzaalvarez70) [reply src/coin/eth/transactionBuilder.ts:141]: That's part of story BGA-14

File: src/coin/eth/utils.ts (hunk 4 above)
Comment (gonzaalvarez70) [reply src/coin/eth/utils.ts:113]: Sure, sounds good.

File: src/coin/eth/transferFundsBuilder.ts (hunk 3 above)
Comment (gonzaalvarez70) [reply src/coin/eth/transferFundsBuilder.ts:10]: With that approach, I'm worried about the token contract address, also for deserialization I think that keeping those transactions (sendMultiSig and sendMultiSigToken) separately would be better. I think that we could think that approach as an extension once we start implementing RSK. What do you think?

Comment (FabianSimtlix) [reply src/coin/eth/transferFundsBuilder.ts:10]: I also think that it's better if we separate the builders. That way we can keep each one focus on a particular contract method. The logic that you mentioned would have more sense in transactionBuilder as an strategy there to choose which builder to use based on the given asset.
Also it relates with next changes where we want to split transactions by its type and payload

Comment (marktoda) [reply src/coin/eth/transferFundsBuilder.ts:10]: > I'm worried about the token contract address, also for deserialization

That's the point of `@bitgo/statics` -- it contains coin name to contract address mappings, which you can use both in ser and deser. 

> That way we can keep each one focus on a particular contract method

The difference between each contract method is quite small, and shouldn't affect any part of the builder other than the serialization of the tx

This idea is also what XTZ are doing: https://github.com/BitGo/bitgo-account-lib/blob/e479643c68251d2d8e25235cc95f6c19eebdb646/src/coin/xtz/transferBuilder.ts#L23

Comment (marktoda) [reply src/coin/eth/transferFundsBuilder.ts:10]: Here is how you can get the contract address from @bitgo/statics:

```
import { coins, Erc20Coin } from '@bitgo/statics';

const omg = coins.get('omg');
if (omg instanceof Erc20Coin) {
  console.log(omg.contractAddress);
}
```

Comment (gonzaalvarez70) [reply src/coin/eth/transferFundsBuilder.ts:10]: What do you think about keeping both TransferBuilders but in TransferTokenBuilder instead of manually setting the tokenContractAddress we create a single transfer method in transactionBuilder that receives a "coin" parameter set as the transaction coin type by default (cgld, rsk, eth).
Based on that parameter we infer the proper transfer builder using what we got in that parameter, and if needed the tokenContractAddress from @bitgo/statics. This way we can keep the structure that makes both transfer builder more maintainable.

However I think that cUSD (the main Celo network token) wasn't added to statics yet, so we would need to make that change first to properly infer the contractAddress.

Comment (marktoda) [reply src/coin/eth/transferFundsBuilder.ts:10]: Feel free to submit PRs into @bitgo/statics -- it is meant to be the canonical source of "coins that BitGo supports".

That idea sounds fine -- I still think there is no need for 2 separate builders which are doing wholly the same thing, but, it's OK

Comment (gonzaalvarez70) [reply src/coin/eth/transferFundsBuilder.ts:10]: The idea of two builders is to handle the different fields that each contract call has (sendMultiSig has a data field and sendMultiSigToken has a tokenContractAddress).
And this way those are more maintainable in case we need to add new functionality to handle those transactions since they are different transaction types. Keeping both in one file would cause a few methods to make two different things because the implementation for signing or decoding is different.
Does that make sense to you?

Comment (marktoda) [reply src/coin/eth/transferFundsBuilder.ts:10]: No, I still disagree, but it's not worth the arguemnt -- just continue on

Comment (gonzaalvarez70) [reply src/coin/eth/transferFundsBuilder.ts:10]: No problem, we can just continue with the initial comment idea (just  one `TransferBuilder ` and added a `coin` field), we just thought that keeping both transferBuilder could be valid even when using the new coin field.

File: src/coin/eth/transferBuilder.ts [hunk 5]
```diff
@@ -68,14 +88,25 @@ export class TransferBuilder {
...
-        new BigNumber(this._amount).toNumber(),
-        this._data,
-        this._expirationTime,
-        this._sequenceId,
-        this.getSignature(),
-      );
+      if (this._tokenContractAddress !== undefined) {
+        return sendMultiSigTokenData(
+          this._toAddress,
+          new BigNumber(this._amount).toNumber(),
```
Comment (marktoda) [inline src/coin/eth/transferBuilder.ts:94]: This is lossy and throws when given a big number. Please keep the amount as either a string or bignumber

File: src/coin/eth/transferBuilder.ts [hunk 6]
```diff
@@ -68,14 +88,25 @@ export class TransferBuilder {
...
+          this._tokenContractAddress,
+          this._expirationTime,
+          this._sequenceId,
+          this.getSignature(),
+        );
+      } else {
+        return sendMultiSigData(
+          this._toAddress,
+          new BigNumber(this._amount).toNumber(),
```
Comment (marktoda) [inline src/coin/eth/transferBuilder.ts:103]: ditto above

File: src/coin/eth/transferBuilder.ts [hunk 7]
```diff
@@ -91,6 +122,41 @@ export class TransferBuilder {
...
   }
 
+  protected getOperationHash(): (string | Buffer)[][] {
+    const operationData = this.getOperationData();
+    return ethUtil.bufferToHex(EthereumAbi.soliditySHA3(...operationData));
+  }
+
+  private getOperationData(): any[][] {
+    let operationData;
+    if (this._tokenContractAddress !== undefined) {
```
Comment (marktoda) [inline src/coin/eth/transferBuilder.ts:132]: Readability Nit: You can save the `@bitgo/statics` coin object instead of just the token address, so that you can just check `if (this.coin.isToken)` or `if (this.coin instanceof Erc20Coin)` here, just for readability.. up to you. This may also be useful for CELO where even the native coins have token addresses

File: src/coin/baseCoin/enum.ts [hunk 8]
```diff
@@ -3,6 +3,7 @@
  */
 export enum TransactionType {
   Send,
+  SendToken,
```
Comment (gonzaalvarez70) [inline src/coin/baseCoin/enum.ts:6]: In my opinion, the new transaction type is not needed anymore. What do you think @marktoda  should we remove it?

File: src/coin/eth/transferBuilder.ts [hunk 9]
```diff
@@ -91,6 +122,41 @@ export class TransferBuilder {
     );
   }
 
+  protected getOperationHash(): (string | Buffer)[][] {
+    const operationData = this.getOperationData();
+    return ethUtil.bufferToHex(EthereumAbi.soliditySHA3(...operationData));
+  }
+
+  private getOperationData(): any[][] {
```
Comment (marktoda) [inline src/coin/eth/transferBuilder.ts:130]: Can you define the return type as something better than `any`? Does `(string | Buffer)[][]` not work anymore? I'd also consider thinking about merging this and `getOperationHash` as this operation data will always be just going straight into `soliditySHA3`

File: src/coin/eth/transferBuilder.ts [hunk 10]
```diff
@@ -91,6 +122,41 @@ export class TransferBuilder {
     );
   }
 
+  protected getOperationHash(): (string | Buffer)[][] {
```
Comment (marktoda) [inline src/coin/eth/transferBuilder.ts:125]: wait how is this return type correct? Doesn't it just return a hex string?

File: src/coin/eth/utils.ts [hunk 11]
```diff
@@ -92,9 +86,33 @@ export function sendMultiSigData(
...
+ * @param {number} value Amount to tranfer
+ * @param {string} tokenContractAddress the address of the erc20 token contract
+ * @param {number} expireTime expiration time for the transaction in seconds
+ * @param {number} sequenceId sequence id
+ * @param {string} signature signature of the call
+ * @returns {string} -- the contract method encoded data
+ */
+export function sendMultiSigTokenData(
+  to: string,
+  value: number,
```
Comment (marktoda) [inline src/coin/eth/utils.ts:107]: lossy - string or bignumber for values please

File: test/unit/coin/eth/transferTokenBuilder.ts [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        .coin('tcusd')
+        .expirationTime(1590078260)
+        .amount(amount)
+        .to(toAddress)
+        .contractSequenceId(2)
+        .key(key);
+      const result = builder.signAndBuild();
+      should.equal(result, testData.SEND_TOKEN_DATA);
+    });
+    it('should fail if a coin is not an ERC20 instance', () => {
```
Comment (marktoda) [inline test/unit/coin/eth/transferTokenBuilder.ts:21]: nit spaces between these it clauses

File: test/unit/coin/eth/transferTokenBuilder.ts [hunk 13]
```diff
@@ -0,0 +1,55 @@
+import should from 'should';
+import { TransferBuilder } from '../../../../src/coin/eth';
+import * as testData from '../../../resources/eth/eth';
+
+describe('Eth send token multi sig builder', function() {
+  describe('should build', () => {
+    const toAddress = '0x7325A3F7d4f9E86AE62Cf742426078C3755730d5';
+    const key = '8CAA00AE63638B0542A304823D66D96FF317A576F692663DB2F85E60FAB2590C';
+    const amount = '0.01';
```
Comment (marktoda) [inline test/unit/coin/eth/transferTokenBuilder.ts:9]: wait, we define amount as big units (i.e. with decimal places?) I expected that we would be using base units (i.e. never decimal places). Where are you converting to base units as I don't see that anywhere above?

Comment (marktoda) [reply test/unit/coin/eth/transferTokenBuilder.ts:9]: Note generally our external APIs accept base units, just to keep things simple. This _could_ be OK, but I have a hunch that the value is not being converted properly

File: test/unit/coin/eth/transferTokenBuilder.ts [hunk 14]
```diff
@@ -0,0 +1,55 @@
+import should from 'should';
+import { TransferBuilder } from '../../../../src/coin/eth';
+import * as testData from '../../../resources/eth/eth';
+
+describe('Eth send token multi sig builder', function() {
```
Comment (marktoda) [inline test/unit/coin/eth/transferTokenBuilder.ts:5]: now that theres only 1 builder, can we just add this to another describe block in `test/transferBuilder`?

File: test/unit/coin/eth/transferTokenBuilder.ts [hunk 15]
```diff
@@ -0,0 +1,55 @@
+import should from 'should';
+import { TransferBuilder } from '../../../../src/coin/eth';
+import * as testData from '../../../resources/eth/eth';
+
+describe('Eth send token multi sig builder', function() {
+  describe('should build', () => {
+    const toAddress = '0x7325A3F7d4f9E86AE62Cf742426078C3755730d5';
+    const key = '8CAA00AE63638B0542A304823D66D96FF317A576F692663DB2F85E60FAB2590C';
```
Comment (marktoda) [inline test/unit/coin/eth/transferTokenBuilder.ts:8]: Please also add tests that test using `xpubs` (as that's what is used in bitgo more often than raw keys)

File: test/unit/coin/eth/transferTokenBuilder.ts [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+        .contractSequenceId(2)
+        .key(key);
+      const result = builder.signAndBuild();
+      should.equal(result, testData.SEND_TOKEN_DATA);
+    });
+    it('should fail if a coin is not an ERC20 instance', () => {
+      should.throws(() => {
+        new TransferBuilder().coin('eth');
+      }, 'There was an error using that coin as a transfer currency');
+    });
```
Comment (marktoda) [inline test/unit/coin/eth/transferTokenBuilder.ts:25]: We can think of some other edge cases to test here, hopefully to avoid finding issues later on. Some ideas:
- a coin that isn't in @bitgo/statics at all
- a coin that is a token in @bitgo/statics but not an erc20 token (https://github.com/BitGo/BitGoJS/blob/11b60b28046f8800f67d02fe235b85c15e449f5d/modules/statics/src/coins.ts#L354)
- invalid key, sequence id, destination, amount (not just missing), expiration
- success with very large amount, sequence id

File: src/coin/baseCoin/enum.ts (hunk 8 above)
Comment (marktoda) [reply src/coin/baseCoin/enum.ts:6]: :+1:

File: src/coin/eth/transferBuilder.ts (hunk 7 above)
Comment (gonzaalvarez70) [reply src/coin/eth/transferBuilder.ts:139]: Sure, I'll go with this option could be useful in the future to have that coin stored.

File: src/coin/eth/transferBuilder.ts (hunk 9 above)
Comment (gonzaalvarez70) [reply src/coin/eth/transferBuilder.ts:137]: Yeah I will change that.

File: src/coin/eth/transferBuilder.ts (hunk 10 above)
Comment (gonzaalvarez70) [reply src/coin/eth/transferBuilder.ts:132]: You are right, this actually returns a string.

File: test/unit/coin/eth/transferTokenBuilder.ts (hunk 12 above)
Comment (gonzaalvarez70) [reply test/unit/coin/eth/transferTokenBuilder.ts:21]: Sure

File: test/unit/coin/eth/transferTokenBuilder.ts (hunk 13 above)
Comment (gonzaalvarez70) [reply test/unit/coin/eth/transferTokenBuilder.ts]: I just changed this, we are not taking numbers with decimal places with the latest fix for BGA-120 as I understand.

File: test/unit/coin/eth/transferTokenBuilder.ts (hunk 14 above)
Comment (gonzaalvarez70) [reply test/unit/coin/eth/transferTokenBuilder.ts:5]: You're right. Done.

File: test/unit/coin/eth/transferTokenBuilder.ts (hunk 15 above)
Comment (gonzaalvarez70) [reply test/unit/coin/eth/transferTokenBuilder.ts:8]: Ok I'll add that.

Comment (gonzaalvarez70) [reply test/unit/coin/eth/transferTokenBuilder.ts:8]: Can we derive a private key from an xpub? Maybe we need to use an xprv. Am I right?

File: test/unit/coin/eth/transferTokenBuilder.ts (hunk 16 above)
Comment (gonzaalvarez70) [reply test/unit/coin/eth/transferTokenBuilder.ts:25]: Ok I'll add some new test cases sounds good.

File: src/coin/eth/transferBuilder.ts (hunk 5 above)
Comment (gonzaalvarez70) [reply src/coin/eth/transferBuilder.ts]: Solved as part of BGA-120

File: src/coin/eth/transferBuilder.ts (hunk 6 above)
Comment (gonzaalvarez70) [reply src/coin/eth/transferBuilder.ts]: Solved as part of BGA-120

File: src/coin/eth/transferBuilder.ts (hunk 10 above)
Comment (marktoda) [reply src/coin/eth/transferBuilder.ts]: I'm just bewildered that it compiled like that before, hmm

File: test/unit/coin/eth/transferTokenBuilder.ts (hunk 15 above)
Comment (marktoda) [reply test/unit/coin/eth/transferTokenBuilder.ts]: ah yep xprv

**Comment by gonzaalvarez70** [comment]
> LGTM, though tests seem broken

We would actually need for https://github.com/BitGo/BitGoJS/pull/753 to be merged and add that new @bitgo/statics reference in order to make the tests work.

File: src/coin/eth/transferBuilder.ts [hunk 17]
```diff
@@ -94,6 +137,47 @@ export class TransferBuilder {
...
+  private getOperationData(): (string | Buffer)[][] {
+    let operationData;
+    if (this._tokenContractAddress !== undefined) {
+      operationData = [
+        ['string', 'address', 'uint', 'address', 'uint', 'uint'],
+        [
+          'ERC20',
+          new ethUtil.BN(ethUtil.stripHexPrefix(this._toAddress), 16),
+          this._amount,
+          new ethUtil.BN(ethUtil.stripHexPrefix(this._tokenContractAddress), 16),
```
Comment (twtaylorbitgo) [inline src/coin/eth/transferBuilder.ts:161]: if the only variation for this is one parameter, you should be writing an inline ternary statement to encapsulate this or just varying that one parameter. this is a lot of extra lines of code here

File: src/coin/eth/transferBuilder.ts [hunk 18]
```diff
@@ -148,20 +217,38 @@ export class TransferBuilder {
...
+    if (!(data.startsWith(sendMultisigMethodId) || data.startsWith(sendMultisigTokenMethodId))) {
       throw new BuildTransactionError(`Invalid transfer bytecode: ${data}`);
     }
-    const splitBytecode = data.split(sendMultisigMethodId);
-    if (splitBytecode.length !== 2) {
-      throw new BuildTransactionError(`Invalid send bytecode: ${data}`);
+    let decoded;
+    if (this.isTokenTransfer(data)) {
+      decoded = this.getRawDecoded(sendMultiSigTokenTypes, this.getBufferedByteCode(sendMultisigTokenMethodId, data));
+      this._tokenContractAddress = ethUtil.bufferToHex(decoded[2]);
```
Comment (twtaylorbitgo) [inline src/coin/eth/transferBuilder.ts:227]: instead of a 2, can you put this in a constant index describing what this field describes? 

https://stackoverflow.com/questions/4117214/using-constants-as-indices-for-javascript-associative-arrays

Comment (gonzaalvarez70) [reply src/coin/eth/transferBuilder.ts:227]: Sounds good, I'll create a new Enum to describe this index.

File: src/coin/eth/transferBuilder.ts (hunk 17 above)
Comment (gonzaalvarez70) [reply src/coin/eth/transferBuilder.ts:161]: The first array of field types is different and the second one has two different fields. I could do it in a more compact way but I thought that this way would help readability.

Comment (gonzaalvarez70) [reply src/coin/eth/transferBuilder.ts:162]: The only way that I can think about is the following. Let me know if it sounds good
`  private getOperationData(): (string | Buffer)[][] {
    const operationData = [
      this.isTokenOperation()
        ? ['string', 'address', 'uint', 'address', 'uint', 'uint']
        : ['string', 'address', 'uint', 'bytes', 'uint', 'uint'],
      [
        this.isTokenOperation() ? 'ERC20' : 'ETHER',
        new ethUtil.BN(ethUtil.stripHexPrefix(this._toAddress), 16),
        this._amount,
        this.isTokenOperation()
          ? new ethUtil.BN(ethUtil.stripHexPrefix(this._tokenContractAddress), 16)
          : new Buffer(ethUtil.stripHexPrefix(this._data) || '', 'hex'),
        this._expirationTime,
        this._sequenceId,
      ],
    ];
    return operationData;
  }

  private isTokenOperation(): boolean {
    return this._tokenContractAddress !== undefined;
  }`

File: src/coin/eth/transferBuilder.ts (hunk 18 above)
Comment (twtaylorbitgo) [reply src/coin/eth/transferBuilder.ts]: thanks

File: src/coin/eth/transferBuilder.ts (hunk 17 above)
Comment (twtaylorbitgo) [reply src/coin/eth/transferBuilder.ts:161]: ahhh I see. the bytes variation. Ok yeah I think your original is way more readable. Gonna dismiss this, thanks
